resolves #53: Highlight today

kstenschke opened the pull request:
this PR additionally contains the following visual suggestions:
-naming colors by their intended use instead of their color, makes it more intuitive/easier to understand and work on code for ui rendering.
-less variation of color contrast and font sizes should make the overall ui appearance more usable and harmonic.
-to highlight the current day i used a light blue color, roughly following the jetbrains darcula theme. Until having support for themes, i recommend defining a color theme to follow as a basic style guide.

Commit:
highlight today (plus suggestions for a clearer ui style: adapted header coloring to overall colors+contrast, renamed colors by meaning/usage (instead of by color) to ease understanding and possibly later theme'ing, reduced+homogenized day label font size for less cluttered ui, added dividers between day labels to improve visual clarity)

Commit:
resolves #53: Highlight today (plus suggestions for a clearer ui style: adapted header coloring to overall colors+contrast, renamed colors by meaning/usage (instead of by color) to ease understanding and possibly later theme'ing, reduced+homogenized day label font size for less cluttered ui, added dividers between day labels to improve visual clarity)

Commit:
Merge remote-tracking branch 'origin/53' into 53

ohwgiles commented:
![53](https://user-images.githubusercontent.com/1444499/50413628-26393a00-0819-11e9-9563-f156db74968b.png)

There seems to be an unrelated bug with the red line marker not on the correct day :man_facepalming: 

Are you using a dark theme? For me (light theme), the dark header looks out of place, and the hour labels are now too pale.

Frohe Weihnachten!

kstenschke commented:
חג שמח and merry christmas :-)
yes, i am using a dark theme (in which the hour labels were barely readable). after not having worked on focal for a while i wasn't sure whether you changed the main colors to using a dark theme in the meanwhile. i now suggested a related feature, see issue #63 
What would be the easiest way to get a light/dark detection merged together with the feature already in the PR? i suggest that i work more on this branch and try to resolve both issues within one PR than, if you agree?

ohwgiles commented:
In general I think it's better to keep issues small and focused, this results in a git history which is easier to understand if necessary. But these issues are close enough (and should hopefully be small enough) that if you'd like to do it in one branch it's totally fine with me

Commit:
week view ui (w/o events): changed redundantly allocated colors into being initially defined and reused, made ui work on light and dark themes

ohwgiles commented on src/week-view.c:
```diff
@@ -67,6 +69,14 @@ enum {
+GdkRGBA *color_fg_50;
+GdkRGBA *color_header_divider;
+GdkRGBA *color_current_time_marker;
+GdkRGBA *color_current_day;
```
In general global variables should be avoided - I suggest making these static or putting them inside a static struct named e.g. `palette` or `colors`. Furthermore, there's no need for these to be pointers which have to be malloc'd later, they can be regular variables. Then `allocate_colors` is unnecessary

kstenschke replied:
i agree, thank you for the hint how to use an attribute here.

ohwgiles commented on src/week-view.c:
```diff
@@ -196,44 +211,91 @@ static void week_view_draw(WeekView* wv, cairo_t* cr)
+	color_current_day = (GdkRGBA *)malloc(sizeof(GdkRGBA));
+}
+
+void init_style(GtkWidget *mainWindow) {
```
I don't like the way `mainWindow` gets involved here. Typically, the best place for this kind of function would be `week_view_class_init`, which gets called only once no matter how many `WeekView`s are created. `WeekView` is also a `GtkWidget` so you should be able to call `gtk_widget_get_style_context` on it.

I do remember reading somewhere that the style is not available until the widget is realized. This is presumably to support different instances having different colors. In this case, I would suggest making two static structures `colors_dark` and `colors_light`, connecting a handler to `WeekView`s `realized` signal, and updating a new pointer in the `WeekView` instance to point either to `colors_dark` or `colors_light` depending on the style context.

kstenschke replied:
i did not find a way to get a WeekView reference that would allow retrieval of the style context within week_view_class_init, therefor moved it into week_view_init first of all.
I suggest using a programmatic initialization instead of static structures, as it allows for a mix of 1. retrievable colors, plus 2. generical calculation of inbetween colors which are missing in style context (unless using alpha values is sufficient for those), and 3. static color values for dark and light colors.

ohwgiles replied:
Right, you cannot get a `WeekView` instance from `week_view_class_init`. What I meant was something more like this:
```
struct WeekViewColors {
  GdkRGBA bg;
  ...
};

static struct WeekViewColors colors_light, colors_dark;

struct _WeekView {
  ...
  WeekViewColors* colors;
  ...
};

...

static void week_view_class_init(WeekViewClass* klass)
{
  ...
  gdk_rgba_parse_color(colors_light.bg, "#fff");
  gdk_rgba_parse_color(colors_dark.bg, "#000");
  ...
}

static void on_realize(WeekView* wv)
{
  ...
  // retrieve gtkstylecontext
  if(is_dark_from_gtkstylecontext)
    wv->colors = &colors_dark;
  else
    wv->colors = &colors_light;
  ...
}
```

This way, the colors are only allocated once even if there are multiple `WeekView` instances, and each instance does the style check and updates its attribute to point to the appropriate static struct

ohwgiles replied:
Actually I think it's also OK to allocate the colours directly on discovering the `GtkStyleContext` theme type, as you currently do. In practice we won't have multiple `WeekView` instances, but if we did it would be nice that they wouldn't all contain a duplicate struct of colours, but just a pointer to a shared struct. So you could consider this a micro-optimisation. Anyway probably best to just prefer whichever approach you find the most readable

ohwgiles commented on src/week-view.c:
```diff
@@ -196,44 +211,91 @@ static void week_view_draw(WeekView* wv, cairo_t* cr)
+		gdk_rgba_parse(color_fg, "#4d4d4d");
+		gdk_rgba_parse(color_fg_50, "#a6a6a6");
+		gdk_rgba_parse(color_current_time_marker, "#ff0000");
+		gdk_rgba_parse(color_current_day, "#356797");
```
I might have to tweak some of these :)

At some point I think it should be possible to retrieve all these directly from the theme instead of hard-coding them manually, but this is OK for now

kstenschke replied:
agreed, the colors can/should be fully generic in a later stage. (note: light/dark theme detection probably is not sufficient w/ e.g. high contrast themes, though i am sceptic about how relevant those are)

ohwgiles commented on src/week-view.h:
```diff
@@ -21,6 +21,9 @@
 G_DECLARE_FINAL_TYPE(WeekView, week_view, FOCAL, WEEK_VIEW, GtkDrawingArea)
 
+void allocate_colors(void);
+void init_style(GtkWidget *mainWindow);
```
Global functions should be prefixed with the module to which they relate `week_view_*`. But I think both these functions can be removed as I wrote above.

kstenschke replied:
i removed them now.

ohwgiles commented on src/week-view.c:
```diff
@@ -196,44 +211,91 @@ static void week_view_draw(WeekView* wv, cairo_t* cr)
+		cairo_move_to(cr, x + 8, wv->y + HEADER_HEIGHT - 14);
+		cairo_set_font_size(cr, 14);
+
+		if (day.month == wv->now.month && day.day == wv->now.day) {
```
Can you just use `wv->now.visible && day.day == wv->now.day` and then there is no need for the `month` variable at all? Otherwise, for completeness, you would also have to check the year, since the same day of the same month reoccurs every few years...

`visible` really means "current span includes today", perhaps a better name is needed

kstenschke replied:
i suggest renaming to wv->now.within_shown_range to make this attribute more "speaking" and more homogenous w/ the other shown_... attributes. see fbfaf13

Commit:
removed global color variables and rel. manual allocation,
moved colors into struct in weekView,
moved colors initialization (independent from mainWindow) into week_view_init,
simplified detection of current day during rendering iteration,
removed unused attribute wv->now.month,
renamed wv->now.visible to wv->now.within_shown_range (more homogenous w/ other shown_... attributes)

ohwgiles commented:
Detection isn't working for me with the latest commit:
```
GTK_THEME=Adwaita-dark ./focal
```
![53-1](https://user-images.githubusercontent.com/1444499/50534440-416fb680-0b45-11e9-9045-60474e33f88d.png)
```
GTK_THEME=Adwaita ./focal
```
![53-2](https://user-images.githubusercontent.com/1444499/50534436-34eb5e00-0b45-11e9-845a-56cbe945e953.png)

I think this is because `week_view_init` is too early to correctly retrieve the style context. Try moving it into a handler connected to the [realize](https://developer.gnome.org/gtk3/stable/GtkWidget.html#GtkWidget-realize) signal?

Commit:
moved color initialization into on_realize handler, evoked by listener on wv.realize

kstenschke commented:
it should work now, looks ok using adwaita default and dark at my place.

ohwgiles commented:
Much better, thanks :)
